src/main/java/appeng/me/cache/NetworkMonitor.java

```diff
@@ -21,283 +21,115 @@
+	protected IMEInventoryHandler getHandler()
 	{
 		switch( this.myChannel )
 		{
 			case ITEMS:
-				return (IMEInventoryHandler<T>) this.myGridCache.getItemInventoryHandler();
+				return this.myGridCache.getItemInventoryHandler();
 			case FLUIDS:
-				return (IMEInventoryHandler<T>) this.myGridCache.getFluidInventoryHandler();
+				return this.myGridCache.getFluidInventoryHandler();
 			default:
 		}
 		return null;
 	}
 
-	private Iterator<Entry<IMEMonitorHandlerReceiver<T>, Object>> getListeners()
-	{
-		return this.listeners.entrySet().iterator();
-	}
-
-	private T monitorDifference( final IAEStack original, final T leftOvers, final boolean extraction, final BaseActionSource src )
-	{
-		final T diff = (T) original.copy();
-
-		if( extraction )
-		{
-			diff.setStackSize( leftOvers == null ? 0 : -leftOvers.getStackSize() );
-		}
-		else if( leftOvers != null )
-		{
-			diff.decStackSize( leftOvers.getStackSize() );
-		}
-
-		if( diff.getStackSize() != 0 )
-		{
-			this.postChangesToListeners( ImmutableList.of( diff ), src );
-		}
-
-		return leftOvers;
-	}
-
-	private void notifyListenersOfChange( final Iterable<T> diff, final BaseActionSource src )
-	{
-		final Iterator<Entry<IMEMonitorHandlerReceiver<T>, Object>> i = this.getListeners();
-
-		while( i.hasNext() )
-		{
-			final Entry<IMEMonitorHandlerReceiver<T>, Object> o = i.next();
-			final IMEMonitorHandlerReceiver<T> receiver = o.getKey();
-			if( receiver.isValid( o.getValue() ) )
-			{
-				receiver.postChange( this, diff, src );
-			}
-			else
-			{
-				i.remove();
-			}
-		}
-	}
-
-	private void postChangesToListeners( final Iterable<T> changes, final BaseActionSource src )
+	@Override
+	protected void postChangesToListeners( final Iterable<T> changes, final BaseActionSource src )
 	{
 		this.postChange( true, changes, src );
 	}
 
-	private void updateCache( Iterable<T> changes, boolean add )
-	{
-		for( final T changedItem : changes )
-		{
-			T difference = changedItem;
-
-			if( !add && changedItem != null )
-			{
-				difference = changedItem.copy();
-				difference.setStackSize( -changedItem.getStackSize() );
-			}
-
-			this.cachedList.add( difference );
-		}
-	}
-
 	protected void postChange( final boolean add, final Iterable<T> changes, final BaseActionSource src )
 	{
-		if( localDepthSemaphore > 0 || GLOBAL_DEPTH.contains( this ) )
+		if( DEPTH.contains( this ) )
 		{
 			return;
 		}
 
-		GLOBAL_DEPTH.push( this );
-		localDepthSemaphore++;
+		DEPTH.push( this );
 
 		this.sendEvent = true;
-
-		this.updateCache( changes, add );
 		this.notifyListenersOfChange( changes, src );
 
+		final IItemList<T> myStorageList = this.getStorageList();
+
 		for( final T changedItem : changes )
 		{
 			T difference = changedItem;
 
 			if( !add && changedItem != null )
 			{
-				difference = changedItem.copy();
-				difference.setStackSize( -changedItem.getStackSize() );
+				( difference = changedItem.copy() ).setStackSize( -changedItem.getStackSize() );
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Extract the assignment out of this expression. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonar.orod.de/coding_rules#rule_key=squid%3AAssignmentInSubExpressionCheck)